package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,355 @@
+
+class MeanSquaredDisplacement(object):
+    r"""Class representing Mean Squared Displacement
+
+    Parameters
+    ----------
+    u : Universe
+        An MDAnalysis :class:`Universe`
+    selection : str
+        An MDAnalysis selection string
+    msd_type : str
+        The dimensions to use for the msd calculation:
+        one of ('xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z')
+    fft : bool
+        Use a fast FFT based algorithm for computation of the MSD
+
+    Returns
+    -------
+    MSD : :class:`MeanSquaredDisplacement`
+        the MSD class
+    
+    """
+    # Attributes
+    # ----------
+    # _dim_fac : int
+    #     dimensionality of the MSD
+    # _dim : arraylike
+    #     array used to slice the trajectory along the xyz axis to acheive the correct dimensionality
+    # _position_array : :class:`np.ndarray`
+    #     positions used to calculate the MSD
+    # N_particles : int
+    #     number of particles 
+
+
+    def __init__(self, u, selection, msd_type='xyz', fft=True):
+
+        #args
+        self.u = u
+        self.selection = selection
+        self.msd_type = msd_type
+        self.fft = fft
+        
+        #local
+        self.dim_fac = 0
+        self._dim = None
+        self.atoms = None
+        self.n_frames = len(self.u.trajectory)
+        self._position_array = None
+
+        #result
+        self.timeseries = None
+        #prep
+        self._prepare()
+
+            
+    def _prepare(self):
+        self.parse_msd_type()
+        self.select_reference_positions()
+        
+    def parse_msd_type(self):
+        r""" Sets up the desired dimensionality of the MSD
+        
+        Parameters
+        ----------
+        self.msd_type : str
+            the desired dimensionality fo the MSD
+        
+        Returns
+        -------
+        self._dim : list
+            array-like used to slice the positions to obtain desired dimensionality
+        self.dim_fac : float
+            dimension factor d of the MSD
+        """
+        
+        if self.msd_type == 'xyz': # full 3d
+            self._dim = [0,1,2]
+            self.dim_fac = 3.0
+            
+        elif self.msd_type == 'xy': # xy
+            self._dim = [0,1]
+            self.dim_fac = 2.0
+
+        elif self.msd_type == 'xz': # xz
+            self._dim = [0,2]
+            self.dim_fac = 2.0
+
+        elif self.msd_type == 'yz': # yz
+            self._dim = [1,2]
+            self.dim_fac = 2.0
+
+        elif self.msd_type == 'x': # x
+            self._dim = [0]
+            self.dim_fac = 1.0
+
+        elif self.msd_type == 'y': # y
+            self._dim = [1]
+            self.dim_fac = 1.0
+
+        elif self.msd_type == 'z': # z
+            self._dim = [2]
+            self.dim_fac = 1.0
+
+        else:
+            raise ValueError('invalid msd_type specified')
+
+    def select_reference_positions(self):
+        #memory transfer required to access coordinate array
+        self.u.transfer_to_memory()
```

Comment:
`transfer_to_memory()` isn't going to be viable for many trajectories, so making it the default approach seems unwise

Reply:
We do need a full array of positions before the main calculation loop can begin. `transfer_to_memory()` gives us access to `.coordinate_array` but obviously is limited by size. The only other option i can think of is tiling out the array in a seperate method attached to `_prepare()` that does it iteratively.